typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,726 @@
+The first one and the most powerful is [immer](https://immerjs.github.io/immer/).
+It is used for immutable state flow.
+Immutable data is not copied but shared in memory.
+It also has helpers for React.
+
+```typescript
+type A = Immutable<
+  {
+    text: string;
+    valid: boolean;
+  }[]
+>;
+
+const a: A = [
+  {
+    text: "valid text",
+    valid: true,
+  },
+  {
+    text: "invalid text",
+    valid: false,
+  },
+];
+
+const nextA = produce(a, (draft) => {
+  draft[1].valid = false;
+  draft.push({ text: "some text", valid: false });
+});
+```
+
+And finally there are three more interesting libraries for working with immutable structures which may be more familiar for Haskell devs: [partial.lenses](https://github.com/calmm-js/partial.lenses), [monocle-ts](https://github.com/gcanti/monocle-ts), and [optics-ts](https://github.com/akheron/optics-ts).
+These are optics libraries.
+Unfortunately the first one doesn't have type bindings.
+The second one support TypeScript and is a partial porting of Scala monocle.
+And the last one is the newest and TypeScript oriented optics library.
+
+### High Order functions and Currying
+
+A lot of languages support High Order functions and TypeScript is not an exception.
+You can use functions as an argument and also return them.
+
+```typescript
+const filterArray = (fn: (elem: string) => boolean, arr: string[]): string[] => {
+  const newArray: string[] = [];
+  for (let i = 0; i < arr.length; i++) {
+    if (fn(arr[i])) {
+      newArray.push(arr[i]);
+    }
+  }
+  return newArray; // or just simple return arr.filter(fn);
+};
+
+const f = (x: string): ((s: number) => boolean) => {
+  return (s: number) => {
+    return s.toString() === x;
+  };
+};
+
+f("5")(5);
+```
+
+As you may see in the last example, we can call function `f` and return function from it, applying in for the new argument.
+So, yes, we can implement some king of currying using it.
+
+```typescript
+type addT = (_: number) => (_: number) => number;
+const add: addT = (l) => (r) => l + r;
+
+type foldArrT = (f: (_: number) => (_: number) => number) => (_: number) => (_: number[]) => number;
+const foldArr: foldArrT = (f) => (z) => (arr) => {
+  let m = z;
+  arr.forEach((elem) => {
+    m = f(m)(elem);
+  });
+  return m;
+};
+
+foldArr(add)(0)([1, 2, 3, 4, 5]);
+```
+
+Unfortunately syntax is weird as you need to always use named parameters.
+But it is still a possible functionality, and you can simply use it in your code.
+
+### Polymorphism
+
+There are different polymorphisms, but we will say only about thee of them: Parametric, Ad hoc, and Row.
+Let's start from the first on.
+Parametric polymorphism allows us to write abstract function or data which will not depend on its type.
+In TypeScript, it is called generics.
+
+```typescript
+const arrLength: <T>(_: T[]) => number = (arr) => arr.length;
+
+arrLength<number>([1, 2, 3]);
+arrLength<string>(["1", "2", "3"]);
```

Comment:
It becomes harder to understand what's going on. E.g. what `arrLength` returns (type or value?), where can I use it?

Reply:
>It becomes harder to understand what's going on. E.g. what arrLength returns (type or value?), where can I use it?

Hm, in the previous part about `currying` there was an example `add: addT`. It show how we can split type and implementation and also the syntax. \
So this is like 
```haskell
arrLength :: [a] -> Int
arrLength l = length l
```
But written smth like `arrLength :: [a] -> Int = l -> length l` (It is funny that students on Haskell course were asking about such syntax) \

> where can I use it

Everywhere? It is just a function (maybe not really useful)